src/HybridSchrodingerFeynmanSimulator.cpp

```diff
@@ -131,116 +134,98 @@ std::map<std::string, std::size_t> HybridSchrodingerFeynmanSimulator::Simulate(u
+        executor.silent_async([&compute_pair, i]() { compute_pair(std::make_pair(0, i)); });
     }
-    omp_destroy_lock(&stacklock);
+    executor.wait_for_all();
+
     root_edge = dd->deserialize<dd::Package::vNode>("slice_" + std::to_string(ndecisions) + "_0.dd", true);
     dd->incRef(root_edge);
 }
+
 void HybridSchrodingerFeynmanSimulator::SimulateHybridAmplitudes(dd::Qubit split_qubit) {
     const auto         ndecisions  = getNDecisions(split_qubit);
     const std::int64_t max_control = 1LL << ndecisions;
 
     const int actuallyUsedThreads = static_cast<std::size_t>(max_control) < nthreads ? static_cast<int>(max_control) : static_cast<int>(nthreads);
-    omp_set_num_threads(actuallyUsedThreads);
-    root_edge = qc::VectorDD::zero;
+    root_edge                     = qc::VectorDD::zero;
 
     const std::int64_t   nslices_on_one_cpu = std::min<std::int64_t>(64, max_control / actuallyUsedThreads);
     const dd::QubitCount nqubits            = getNumberOfQubits();
 
     std::vector<std::vector<std::complex<dd::fp>>> amplitudes(actuallyUsedThreads, std::vector<std::complex<dd::fp>>(1u << nqubits));
 
-#pragma omp parallel for schedule(dynamic, 1) // NOLINT(openmp-use-default-none)
-    for (std::int64_t control = 0; control < max_control; control += nslices_on_one_cpu) {
-        const auto                         current_thread    = omp_get_thread_num();
-        std::vector<std::complex<dd::fp>>& thread_amplitudes = amplitudes.at(current_thread);
-
-        for (std::int64_t local_control = 0; local_control < nslices_on_one_cpu; local_control++) {
-            const std::int64_t           total_control = control + local_control;
-            std::unique_ptr<dd::Package> slice_dd      = std::make_unique<dd::Package>(getNumberOfQubits());
-            auto                         result        = SimulateSlicing(slice_dd, split_qubit, total_control);
-            slice_dd->addAmplitudes(result, thread_amplitudes, nqubits);
-        }
+    tf::Executor executor;
+    for (std::int64_t control = 0, i = 0; control < max_control; control += nslices_on_one_cpu, i++) {
+        executor.silent_async([this, i, &amplitudes, nslices_on_one_cpu, control, nqubits, split_qubit]() {
+            const auto                         current_thread    = i;
+            std::vector<std::complex<dd::fp>>& thread_amplitudes = amplitudes.at(current_thread);
+
+            for (std::int64_t local_control = 0; local_control < nslices_on_one_cpu; local_control++) {
+                const std::int64_t           total_control = control + local_control;
+                std::unique_ptr<dd::Package> slice_dd      = std::make_unique<dd::Package>(getNumberOfQubits());
+                auto                         result        = SimulateSlicing(slice_dd, split_qubit, total_control);
+                slice_dd->addAmplitudes(result, thread_amplitudes, nqubits);
+            }
+        });
     }
+    executor.wait_for_all();
```

Comment:
technically, task flow supports a `for_each` construct, see: https://taskflow.github.io/taskflow/index.html#QuickStartLeverageStandardParallelAlgorithms
which might be a good fit here.

Reply:
Mh... We would still need to construct some iterable with the controls first and `for_each` does a similar construction as the current code underneath. This seems to be something to keep in mind for change as we get a better grasp on taskflow.